### PR TITLE
Ensure one-shot wrappers release their delegates

### PR DESCRIPTION
**Pull request by DaveCTurner**
We recently adjusted `RunOnce`, `Releasables#releaseOnce` and `ActionListener#notifyOnce` so that once they have fired they drop the now-unnecessary reference to the delegate. This commit introduces tests to verify that this reference does genuinely become unreachable (i.e. available for garbage collection) as expected.

It also fixes a bug in `ActionListener#notifyOnce` which caused us to unexpectedly retain a reference to the delegate 🤦

Relates #92452
Relates #92507
Relates #92537

### CODE REVIEW

**Comment by DaveCTurner** [comment]
This works reliably on my machine but my main question to reviewers is whether you think this will work everywhere our test suite expects to run. In particular, explicitly triggering the GC is not guaranteed to work in all cases (but it seems that it does in the test suite, right?)

**Comment by grcevski** [comment]
Explicitly triggering the GC doesn't guarantee that the references will be marked as unreachable. I think it will work for us, but we need to be careful. MXBean.gc() is equivalent to System.gc() as far as I know. It will cause a GC on G1, but in most cases it will simply be a new space collection, which means it will only remove the references of the recently allocated objects.

If there was high allocation pressure the collector may decide to tenure the object, e.g. the object survives at least one minor GC and it's tenured. In this scenario, the only way the reference will be cleaned up for certain is if we cause a full collect.

The JDK internal tests have this `WhiteBox` helper class that has a very useful fullGC() method, but we can't use it.

File: test/framework/src/test/java/org/elasticsearch/test/ReachabilityCheckerTests.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.test;
+
+import org.hamcrest.Matchers;
+
+public class ReachabilityCheckerTests extends ESTestCase {
+
+    public void testSuccess() {
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply test/framework/src/test/java/org/elasticsearch/test/ReachabilityCheckerTests.java:15]: I put a GC call in the `ReachabilityChecker` constructor; I'd like this to be more generally useful so I think requiring it in a `@Before` method is not going to work.

File: server/src/test/java/org/elasticsearch/common/ReleasablesTests.java [hunk 2]
```diff
@@ -24,4 +25,13 @@ public void testReleaseOnce() {
         releasable.close();
         assertEquals(1, count.get());
     }
+
+    public void testReleaseOnceReleasesDelegate() {
+        final var reachabilityChecker = new ReachabilityChecker();
+        final var releaseOnce = Releasables.releaseOnce(reachabilityChecker.register(() -> logger.info("test")));
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/common/ReleasablesTests.java]: Fair point, a no-op method reference seems to work.

File: server/src/test/java/org/elasticsearch/action/StepListenerTests.java [hunk 3]
```diff
@@ -164,4 +165,23 @@ public void testThenCombine() throws Exception {
             assertThat(exception.getMessage(), containsString("simulated"));
         }
     }
+
+    public void testAddedListenersReleasedOnCompletion() {
+        final StepListener<Void> step = new StepListener<>();
+        final ReachabilityChecker reachabilityChecker = new ReachabilityChecker();
+
+        for (int i = between(1, 3); i > 0; i--) {
+            step.addListener(reachabilityChecker.register(ActionListener.wrap(() -> {})));
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/action/StepListenerTests.java:174]: If it needs that much ceremony to be stable then this is not going to work in general. I wonder, would it work to do the same thing that the `G1OverLimitStrategy` does?

[the remark this answers is not in the document]
Comment (grcevski) [reply server/src/test/java/org/elasticsearch/action/StepListenerTests.java:174]: Yeah, I think if we can somehow reuse that, we are definitely more certain the references will be properly checked for liveness. I also think the 3 action listeners are not a lot of garbage, it won't be a problem if we called System.gc() on creation of the ReachabilityChecker. It's fine as-is with the GC-up-front change.

**Review by grcevski — APPROVED** [review]
LGTM!

File: server/src/test/java/org/elasticsearch/action/StepListenerTests.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/action/StepListenerTests.java:174]: Ok, I added a short comment to the class indicating that there is some possible flakiness here. We can revisit this if it turns out to be bad.

**Pull request merged**